internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -347,11 +354,18 @@ func NewReconciler(mgr manager.Manager, of resource.CompositeKind, opts ...Recon
 		client:       kube,
 		newComposite: nc,
 
-		composition: composition{
-			CompositionFetcher: NewAPICompositionFetcher(kube),
-			CompositionValidatorFunc: func(in *v1.Composition) error {
-				return in.Validate().ToAggregate()
-			},
+		revision: revision{
+			CompositionRevisionFetcher: NewAPIRevisionFetcher(resource.ClientApplicator{Client: kube, Applicator: resource.NewAPIPatchingApplicator(kube)}),
+			CompositionRevisionValidator: CompositionRevisionValidatorFn(func(rev *v1.CompositionRevision) error {
+				// TODO(negz): Presumably this validation will eventually be
+				// removed in favor of the new Composition validation
+				// webhook.
+				// This is the last remaining use of conv.FromRevisionSpec -
+				// we can stop generating that once this is removed.
```

Comment:
@phisco Does this XR reconciler need to keep validating Compositions?

I'm guessing we do for now, assuming that the new validation webhook will be alpha and off by default.

Reply:
I would leave it, users can always delete the validating webhook, and validating webhooks do not run on existing resources upon upgrades, obviously it should never happen that a valid composition becomes invalid after an upgrade, nonetheless we should catch it as soon as possible.

As per your suggestion [here](https://github.com/crossplane/crossplane/issues/3748) webhooks are now enabled by default, given that the validation is also done in the reconciliation loop, it didn't make sense to split the `webhooks.enabled` value, so both XRD and Compositions will be enabled. The logical validation is not going to be behind feature flag, while future iterations using schemas will be, see https://github.com/crossplane/crossplane/pull/3937.

Reply:
I'll leave it in place for this PR.

In a future PR I'd like to move it into the existing Composition controller, and have that controller set some status condition for invalid Compositions. I would prefer this because it:

* Could be used to ensure we never create a `CompositionRevision` for an invalid `Composition`, thus ensuring no-one can actually use an invalid `Composition`.
* Simplifies the XR controller.
* Avoids us having to convert from `Composition` -> `CompositionRevision` -> `Composition` to run this validation logic.

Does that sound reasonable? If so I may try do that before the freeze on Tuesday.

Reply:
Sounds good to me! In the long run, once the schema-aware validation is promoted to GA, we can maybe think of running that too to compute the validity of a Composition.